python/cudf/cudf/io/parquet.py

```diff
@@ -791,27 +911,105 @@ def __init__(
+        self.max_file_size = max_file_size
+        if max_file_size is not None:
+            if file_name_prefix is None:
+                raise ValueError(
+                    "file_name_prefix cannot be None if max_file_size is "
+                    "passed"
+                )
+            self.max_file_size = _parse_bytes(max_file_size)
+
+        self._file_sizes: Dict[str, int] = {}
 
     @_cudf_nvtx_annotate
     def write_table(self, df):
         """
         Write a dataframe to the file/dataset
         """
-        (
-            paths,
-            metadata_file_paths,
-            grouped_df,
-            offsets,
-            self.filename,
-        ) = _get_partitioned(
-            df,
-            self.path,
-            self.partition_cols,
+        (part_names, grouped_df, part_offsets,) = _get_groups_and_offsets(
+            df=df,
+            partition_cols=self.partition_cols,
             preserve_index=self.common_args["index"],
-            filename=self.filename,
         )
+        fs = ioutils._ensure_filesystem(None, self.path)
+        fs.mkdirs(self.path, exist_ok=True)
+
+        full_paths = []
+        metadata_file_paths = []
+        full_offsets = [0]
+
+        for idx, keys in enumerate(part_names.itertuples(index=False)):
+            subdir = fs.sep.join(
+                [
+                    f"{name}={val}"
+                    for name, val in zip(self.partition_cols, keys)
+                ]
+            )
+            prefix = fs.sep.join([self.path, subdir])
+            fs.mkdirs(prefix, exist_ok=True)
+            current_offset = (part_offsets[idx], part_offsets[idx + 1])
+            num_chunks = 1
+            if self.max_file_size is not None:
+                start, end = current_offset
+                sliced_df = grouped_df[start:end]
+
+                current_file_size = get_estimated_file_size(sliced_df)
+                if current_file_size > self.max_file_size:
+                    parts = int(current_file_size // self.max_file_size)
+                    new_offsets = list(
+                        range(start, end, int((end - start) / parts))
+                    )[1:]
+                    new_offsets.append(end)
+                    num_chunks = len(new_offsets)
+                    full_offsets.extend(new_offsets)
+                else:
+                    full_offsets.append(end)
+
+                curr_file_num = 0
+                while num_chunks > 0:
+                    new_file_name = f"{self.filename}_{curr_file_num}.parquet"
+                    new_full_path = fs.sep.join([prefix, new_file_name])
+                    while (
+                        new_full_path in self._file_sizes
+                        and (
+                            self._file_sizes[new_full_path]
+                            + (current_file_size / num_chunks)
+                        )
+                        > self.max_file_size
+                    ):
```

Comment:
This is the only section I found a little hard to follow. I understand we need a new filepath for each chunk, but I'm having trouble seeing the conditions under which we'd already find the filepath inside `self._file_sizes`. Could you help me out @galipremsagar ?

Reply:
So when we call `write_table` from the same instance on two different data frames, it could be possible that a file name is already taken by a part of the first dataframe, hence we need this check for the second/future data frame to pick a new file name that doesn't exist in `self._file_sizes`

Reply:
Added the same in the comment